File: ProfileManager.py
import json
import os
from PySide6.QtCore import QObject, Signal, Slot, Property

class ProfileManager(QObject):
    """Manages profile and bag data with JSON file persistence"""
    
    profilesChanged = Signal()
    activeProfileChanged = Signal()
    
    def __init__(self):
        super().__init__()
        self._profiles_file = os.path.join(os.path.dirname(__file__), "profiles.json")
        self._profiles_data = self._load_profiles()
        self._active_profile = self._profiles_data.get("active_profile", "Guest")
    
    def _load_profiles(self):
        """Load profiles from JSON file"""
        if os.path.exists(self._profiles_file):
            try:
                with open(self._profiles_file, 'r') as f:
                    return json.load(f)
            except Exception as e:
                print(f"⚠️ Error loading profiles: {e}")
                return self._get_default_data()
        return self._get_default_data()
    
    def _get_default_data(self):
        """Return default profile structure"""
        return {
            "active_profile": "Guest",
            "profiles": ["Guest"],
            "bags": {
                "Guest": {
                    "Default Set": {
                        "Driver": 10.5,
                        "3 Wood": 15.0,
                        "5 Wood": 18.0,
                        "3 Hybrid": 19.0,
                        "4 Iron": 21.0,
                        "5 Iron": 24.0,
                        "6 Iron": 28.0,
                        "7 Iron": 34.0,
                        "8 Iron": 38.0,
                        "9 Iron": 42.0,
                        "PW": 46.0,
                        "GW": 50.0,
                        "SW": 56.0,
                        "LW": 60.0
                    }
                }
            },
            "active_presets": {
                "Guest": "Default Set"
            }
        }
    
    def _save_profiles(self):
        """Save profiles to JSON file"""
        try:
            with open(self._profiles_file, 'w') as f:
                json.dump(self._profiles_data, f, indent=2)
            print(f"✅ Profiles saved to {self._profiles_file}")
        except Exception as e:
            print(f"⚠️ Error saving profiles: {e}")
    
    @Slot(str, result=str)
    def getProfilesJson(self, key):
        """Get a JSON string for QML"""
        if key == "profiles":
            return json.dumps(self._profiles_data.get("profiles", []))
        elif key == "bags":
            return json.dumps(self._profiles_data.get("bags", {}))
        elif key == "active_presets":
            return json.dumps(self._profiles_data.get("active_presets", {}))
        return "[]"
    
    @Slot(str, str)
    def saveProfilesJson(self, key, json_str):
        """Save JSON data from QML"""
        try:
            data = json.loads(json_str)
            self._profiles_data[key] = data
            self._save_profiles()
            self.profilesChanged.emit()
        except Exception as e:
            print(f"⚠️ Error saving {key}: {e}")
    
    @Slot(str)
    def setActiveProfile(self, profile_name):
        """Set the active profile"""
        self._active_profile = profile_name
        self._profiles_data["active_profile"] = profile_name
        self._save_profiles()
        self.activeProfileChanged.emit()
    
    @Slot(result=str)
    def getActiveProfile(self):
        """Get the active profile name"""
        return self._active_profile
    
    @Slot(str)
    def createProfile(self, profile_name):
        """Create a new profile with default bag"""
        if profile_name and profile_name not in self._profiles_data["profiles"]:
            self._profiles_data["profiles"].append(profile_name)
            
            # Copy default bag structure
            default_bag = self._get_default_data()["bags"]["Guest"]["Default Set"]
            self._profiles_data["bags"][profile_name] = {
                "Default Set": default_bag.copy()
            }
            self._profiles_data["active_presets"][profile_name] = "Default Set"
            
            self._save_profiles()
            self.profilesChanged.emit()
    
    @Slot(str)
    def deleteProfile(self, profile_name):
        """Delete a profile"""
        if profile_name in self._profiles_data["profiles"] and profile_name != "Guest":
            self._profiles_data["profiles"].remove(profile_name)
            
            # Remove associated data
            if profile_name in self._profiles_data["bags"]:
                del self._profiles_data["bags"][profile_name]
            if profile_name in self._profiles_data["active_presets"]:
                del self._profiles_data["active_presets"][profile_name]
            
            # If deleting active profile, switch to Guest
            if self._active_profile == profile_name:
                self._active_profile = "Guest"
                self._profiles_data["active_profile"] = "Guest"
            
            self._save_profiles()
            self.profilesChanged.emit()
            self.activeProfileChanged.emit()
    
    @Slot(str, str, str)
    def saveBagPreset(self, profile_name, preset_name, clubs_json):
        """Save a bag preset for a profile"""
        try:
            clubs = json.loads(clubs_json)
            
            if profile_name not in self._profiles_data["bags"]:
                self._profiles_data["bags"][profile_name] = {}
            
            self._profiles_data["bags"][profile_name][preset_name] = clubs
            self._save_profiles()
            self.profilesChanged.emit()
        except Exception as e:
            print(f"⚠️ Error saving bag preset: {e}")
    
    @Slot(str, str, result=str)
    def getBagPreset(self, profile_name, preset_name):
        """Get a bag preset for a profile"""
        try:
            if profile_name in self._profiles_data["bags"]:
                if preset_name in self._profiles_data["bags"][profile_name]:
                    return json.dumps(self._profiles_data["bags"][profile_name][preset_name])
        except Exception as e:
            print(f"⚠️ Error getting bag preset: {e}")
        
        # Return default
        return json.dumps(self._get_default_data()["bags"]["Guest"]["Default Set"])
    
    @Slot(str, str)
    def setActivePreset(self, profile_name, preset_name):
        """Set the active preset for a profile"""
        self._profiles_data["active_presets"][profile_name] = preset_name
        self._save_profiles()

File: src/ProfileManager.cpp
#include "ProfileManager.h"
#include <QFile>
#include <QJsonDocument>
#include <QJsonArray>
#include <QDebug>
#include <QDir>

ProfileManager::ProfileManager(QObject *parent)
    : QObject(parent)
{
    m_profilesFile = QDir::currentPath() + "/profiles.json";
    loadProfiles();
    m_activeProfile = m_profilesData["active_profile"].toString();
}

void ProfileManager::loadProfiles()
{
    QFile file(m_profilesFile);

    if (file.exists() && file.open(QIODevice::ReadOnly)) {
        QByteArray data = file.readAll();
        file.close();

        QJsonDocument doc = QJsonDocument::fromJson(data);
        if (!doc.isNull() && doc.isObject()) {
            m_profilesData = doc.object();
            qDebug() << "Profiles loaded from" << m_profilesFile;
            return;
        }
    }

    // Use default data if file doesn't exist or is invalid
    m_profilesData = getDefaultData();
    qDebug() << "Using default profile data";
}

void ProfileManager::saveProfiles()
{
    QFile file(m_profilesFile);

    if (file.open(QIODevice::WriteOnly)) {
        QJsonDocument doc(m_profilesData);
        file.write(doc.toJson(QJsonDocument::Indented));
        file.close();
        qDebug() << "Profiles saved to" << m_profilesFile;
    } else {
        qWarning() << "Failed to save profiles to" << m_profilesFile;
    }
}

QJsonObject ProfileManager::getDefaultData()
{
    QJsonObject data;
    data["active_profile"] = "";
    data["profiles"] = QJsonArray();
    data["bags"] = QJsonObject();
    data["active_presets"] = QJsonObject();
    return data;
}

QJsonObject ProfileManager::getDefaultBag()
{
    QJsonObject bag;
    bag["Driver"] = 10.5;
    bag["3 Wood"] = 15.0;
    bag["5 Wood"] = 18.0;
    bag["3 Hybrid"] = 19.0;
    bag["4 Iron"] = 21.0;
    bag["5 Iron"] = 24.0;
    bag["6 Iron"] = 28.0;
    bag["7 Iron"] = 34.0;
    bag["8 Iron"] = 38.0;
    bag["9 Iron"] = 42.0;
    bag["PW"] = 46.0;
    bag["GW"] = 50.0;
    bag["SW"] = 56.0;
    bag["LW"] = 60.0;
    return bag;
}

QString ProfileManager::getProfilesJson(const QString &key)
{
    QJsonDocument doc;

    if (key == "profiles") {
        doc.setArray(m_profilesData["profiles"].toArray());
    } else if (key == "bags") {
        doc.setObject(m_profilesData["bags"].toObject());
    } else if (key == "active_presets") {
        doc.setObject(m_profilesData["active_presets"].toObject());
    } else {
        doc.setArray(QJsonArray());
    }

    return QString::fromUtf8(doc.toJson(QJsonDocument::Compact));
}

void ProfileManager::saveProfilesJson(const QString &key, const QString &jsonStr)
{
    QJsonDocument doc = QJsonDocument::fromJson(jsonStr.toUtf8());

    if (!doc.isNull()) {
        if (doc.isArray()) {
            m_profilesData[key] = doc.array();
        } else if (doc.isObject()) {
            m_profilesData[key] = doc.object();
        }

        saveProfiles();
        emit profilesChanged();
    } else {
        qWarning() << "Invalid JSON for key:" << key;
    }
}

void ProfileManager::setActiveProfile(const QString &profileName)
{
    m_activeProfile = profileName;
    m_profilesData["active_profile"] = profileName;
    saveProfiles();
    emit activeProfileChanged();
}

QString ProfileManager::getActiveProfile()
{
    return m_activeProfile;
}

void ProfileManager::createProfile(const QString &profileName)
{
    if (profileName.isEmpty()) {
        return;
    }

    QJsonArray profiles = m_profilesData["profiles"].toArray();

    // Check if profile already exists
    for (const QJsonValue &value : profiles) {
        if (value.toString() == profileName) {
            qDebug() << "Profile already exists:" << profileName;
            return;
        }
    }

    // Add new profile
    profiles.append(profileName);
    m_profilesData["profiles"] = profiles;

    // Create default bag
    QJsonObject bags = m_profilesData["bags"].toObject();
    QJsonObject profileBags;
    profileBags["Default Set"] = getDefaultBag();
    bags[profileName] = profileBags;
    m_profilesData["bags"] = bags;

    // Set default active preset
    QJsonObject activePresets = m_profilesData["active_presets"].toObject();
    activePresets[profileName] = "Default Set";
    m_profilesData["active_presets"] = activePresets;

    saveProfiles();
    emit profilesChanged();

    qDebug() << "Profile created:" << profileName;
}

void ProfileManager::deleteProfile(const QString &profileName)
{
    QJsonArray profiles = m_profilesData["profiles"].toArray();

    // Remove profile from array
    for (int i = 0; i < profiles.size(); ++i) {
        if (profiles[i].toString() == profileName) {
            profiles.removeAt(i);
            break;
        }
    }
    m_profilesData["profiles"] = profiles;

    // Remove associated data
    QJsonObject bags = m_profilesData["bags"].toObject();
    bags.remove(profileName);
    m_profilesData["bags"] = bags;

    QJsonObject activePresets = m_profilesData["active_presets"].toObject();
    activePresets.remove(profileName);
    m_profilesData["active_presets"] = activePresets;

    // If deleting active profile, switch to first available or empty
    if (m_activeProfile == profileName) {
        if (profiles.size() > 0) {
            m_activeProfile = profiles[0].toString();
        } else {
            m_activeProfile = "";
        }
        m_profilesData["active_profile"] = m_activeProfile;
    }

    saveProfiles();
    emit profilesChanged();
    emit activeProfileChanged();

    qDebug() << "Profile deleted:" << profileName;
}

void ProfileManager::saveBagPreset(const QString &profileName, const QString &presetName, const QString &clubsJson)
{
    QJsonDocument doc = QJsonDocument::fromJson(clubsJson.toUtf8());

    if (!doc.isNull() && doc.isObject()) {
        QJsonObject bags = m_profilesData["bags"].toObject();
        QJsonObject profileBags = bags[profileName].toObject();
        profileBags[presetName] = doc.object();
        bags[profileName] = profileBags;
        m_profilesData["bags"] = bags;

        saveProfiles();
        emit profilesChanged();

        qDebug() << "Bag preset saved:" << profileName << "/" << presetName;
    } else {
        qWarning() << "Invalid clubs JSON for bag preset";
    }
}

QString ProfileManager::getBagPreset(const QString &profileName, const QString &presetName)
{
    QJsonObject bags = m_profilesData["bags"].toObject();

    if (bags.contains(profileName)) {
        QJsonObject profileBags = bags[profileName].toObject();

        if (profileBags.contains(presetName)) {
            QJsonDocument doc(profileBags[presetName].toObject());
            return QString::fromUtf8(doc.toJson(QJsonDocument::Compact));
        }
    }

    // Return default bag if not found
    QJsonDocument doc(getDefaultBag());
    return QString::fromUtf8(doc.toJson(QJsonDocument::Compact));
}

void ProfileManager::setActivePreset(const QString &profileName, const QString &presetName)
{
    QJsonObject activePresets = m_profilesData["active_presets"].toObject();
    activePresets[profileName] = presetName;
    m_profilesData["active_presets"] = activePresets;
    saveProfiles();
}
